src/internal/ReplaySubject.ts

```diff
@@ -70,15 +70,36 @@ export class ReplaySubject<T> extends Subject<T> {
     this._throwIfClosed();
     this._trimBuffer();
 
-    const subscription = this._innerSubscribe(subscriber);
-
     const { _infiniteTimeWindow, _buffer } = this;
-    // We use a copy here, so reentrant code does not mutate our array while we're
-    // emitting it to a new subscriber.
-    const copy = _buffer.slice();
-    for (let i = 0; i < copy.length && !subscriber.closed; i += _infiniteTimeWindow ? 1 : 2) {
-      subscriber.next(copy[i] as T);
+
+    // Replay the current values, but also the ones that are re-entered synchronously while emitting them.
+    const syncBuffer = _buffer.slice();
+    let isInitializing = true;
+
+    // This subscriber prevents new values from being emitted before the current values are all flushed.
+    const syncSubscriber = new SafeSubscriber<T>({
+      next(value) {
+        if (isInitializing) {
+          syncBuffer.push(value);
```

Comment:
if there is a window interval, then the buffer has a different shape (see ReplaySubject.next). this needs to be maintained. you could also just use a new array and replay this after the buffer slice was replayed. then you dont need to rely on the buffer shape from next.

Reply:
You are right, this won't work with infiniteTimeWindow, this needs updating.